File: scrapers/betfair/cdp_url.py
from urllib.parse import urlsplit

_LOCAL_HOSTS = {"127.0.0.1", "localhost", "::1"}


def normalize_cdp_base_url(value):
    if value is None:
        return ""

    if not isinstance(value, str):
        return None

    normalized = value.strip().rstrip("/")

    if not normalized:
        return ""

    try:
        parsed = urlsplit(normalized)
        port = parsed.port
    except (TypeError, ValueError):
        return None

    if (
        parsed.scheme != "http"
        or parsed.username is not None
        or parsed.password is not None
        or parsed.hostname not in _LOCAL_HOSTS
        or port is None
        or not (1 <= port <= 65535)
        or parsed.path not in ("", "/")
        or parsed.query
        or parsed.fragment
    ):
        return None

    return normalized


def require_cdp_base_url(value):
    normalized = normalize_cdp_base_url(value)

    if normalized == "":
        raise ValueError("cdp_url_required")

    if normalized is None:
        raise ValueError("cdp_url_invalid")

    return normalized

File: backend/src/utils/cdpUrl.js
const LOCAL_CDP_HOSTS = new Set([
    '127.0.0.1',
    'localhost',
    '::1'
]);

function normalizedHostname(hostname) {
    return String(hostname || '').replace(/^\[|\]$/g, '').toLowerCase();
}

export function normalizeCdpBaseUrl(value) {
    if (value === undefined || value === null) {
        return '';
    }

    if (typeof value !== 'string') {
        return null;
    }

    const normalized = value.trim().replace(/\/+$/, '');

    if (!normalized) {
        return '';
    }

    let parsed;
    try {
        parsed = new URL(normalized);
    } catch (_) {
        return null;
    }

    const port = Number(parsed.port);
    const pathname = parsed.pathname || '/';

    if (
        parsed.protocol !== 'http:' ||
        parsed.username ||
        parsed.password ||
        parsed.search ||
        parsed.hash ||
        pathname !== '/' ||
        !LOCAL_CDP_HOSTS.has(normalizedHostname(parsed.hostname)) ||
        !parsed.port ||
        !Number.isInteger(port) ||
        port < 1 ||
        port > 65535
    ) {
        return null;
    }

    return normalized;
}

export function validateCdpBaseUrl(value) {
    const normalized = normalizeCdpBaseUrl(value);
    return typeof normalized === 'string' && normalized.length > 0;
}

export function buildCdpVersionUrl(value) {
    const normalized = normalizeCdpBaseUrl(value);
    return normalized ? `${normalized}/json/version` : null;
}

export function classifyCdpBaseUrl(value) {
    const normalized = normalizeCdpBaseUrl(value);

    if (normalized === '') {
        return {
            ok: false,
            code: 'cdp_url_required',
            value: ''
        };
    }

    if (normalized === null) {
        return {
            ok: false,
            code: 'cdp_url_invalid',
            value: ''
        };
    }

    return {
        ok: true,
        code: null,
        value: normalized
    };
}
